lib/src/serializable.dart

```diff
@@ -0,0 +1,224 @@
+      _moduleRegistrations[module.serializableKey].dispose();
+      _moduleRegistrations.remove(module.serializableKey);
+    }
+  }
+
+  void registerModule(SerializableModule module) {
+    if (module.serializableKey != null) {
+      final registration = new _ModuleRegistration(module);
+      _registerForLifecycleEvents(registration);
+      _moduleRegistrations[module.serializableKey] = registration;
+    } else {
+      _logger.warning('Unable to serialize module without serializableKey');
+    }
+  }
+
+  void _handleApiCall(Map apiCall) {
+    String module = apiCall['module'];
+    String method = apiCall['method'];
+    Object data = apiCall['data'];
+
+    SerializableModule targetModule = _moduleRegistrations[module].module;
+
+    if (targetModule != null) {
+      _deserializeAndCall(targetModule, method, data);
+    }
+  }
+
+  void _deserializeAndCall(
+      SerializableModule module, String method, List data) {
+    InstanceMirror apiMirror = reflect(module.api);
+
+    if (apiMirror != null) {
+      ClassMirror classMirror = apiMirror.type;
+
+      if (ClassMirror != null) {
+        MethodMirror apiMethodMirror =
+            classMirror.declarations[new Symbol(method)];
+
+        if (apiMethodMirror == null) {
+          _logger.warning(
+              'Method $method does not exist on ${module.serializableKey}\' module\'s API');
+          return;
+        }
+
+        // Check here that the position args in data match the expected params of the method being called
+        if (apiMethodMirror.parameters.length == data.length) {
+          for (var i = 0; i < apiMethodMirror.parameters.length; i++) {
+            var param = apiMethodMirror.parameters[i];
+
+            if (data[i] is Map && param.type.reflectedType != Map) {
+              ClassMirror paramClassMirror =
+                  reflectClass(param.type.reflectedType);
+
+              // Paramter type must implement fromJson name constructor that takes a Map
+              try {
+                var instance = paramClassMirror
+                    .newInstance(new Symbol('fromJson'), [data[i]]).reflectee;
+                data[i] = instance;
+              } on NoSuchMethodError {
+                _logger.warning(
+                    '${paramClassMirror.simpleName.toString()} does not implement fromJson named constructor');
+                return;
+              }
+            }
+          }
+
+          if (apiMirror != null) {
+            try {
+              apiMirror.invoke(new Symbol(method), data);
+            } catch (e) {
+              _logger.severe(
+                  'Unable to call $method on ${module.serializableKey} w_module, ${e.toString()}');
+            }
+          }
+        } else {
+          _logger.warning(
+              'Unable to call api method $method in ${module.serializableKey} w_module, mismatched params');
+        }
+      }
+    }
+  }
+
+  void _registerForAllEvents(SerializableModule module) {
```

Comment:
you'll want to pass in the `_ModuleRegistration` instance here, too, so that you can have it manage the event subscriptions below